v-next/hardhat/src/internal/cli/main.ts

```diff
@@ -265,3 +269,192 @@ function getTaskFromCliArguments(
 
   return task;
 }
+
+function parseTaskArguments(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+): Record<string, any> {
+  const taskArguments: Record<string, unknown> = {};
+
+  parseNamedParameters(cliArguments, usedCliArguments, task, taskArguments);
+
+  parsePositionalAndVariadicParameters(
+    cliArguments,
+    usedCliArguments,
+    task,
+    taskArguments,
+  );
+
+  usedCliArguments.forEach((arg, index) => {
+    if (!arg) {
+      throw new HardhatError(HardhatError.ERRORS.ARGUMENTS.UNUSED_ARGUMENT, {
+        value: cliArguments[index],
+      });
+    }
+  });
+
+  return taskArguments;
+}
+
+function parseNamedParameters(
+  cliArguments: string[],
+  usedCliArguments: boolean[],
+  task: Task,
+  taskArguments: Record<string, any>,
+) {
+  for (let i = 0; i < cliArguments.length; i++) {
+    if (usedCliArguments[i]) {
+      continue;
+    }
+
+    if (cliArguments[i] === "--") {
+      // A standalone '--' is ok because it is used to separate CLI tool arguments from script arguments,
+      // ensuring the tool passes subsequent options directly to the script.
+      continue;
+    }
+
+    const arg = cliArguments[i];
+
+    if (arg.startsWith("--") === false) {
+      continue;
+    }
+
+    const paramName = kebabToCamelCase(arg.substring(2));
+    const paramInfo = task.namedParameters.get(paramName);
+
+    if (paramInfo === undefined) {
+      throw new HardhatError(
+        HardhatError.ERRORS.ARGUMENTS.UNRECOGNIZED_NAMED_PARAM,
+        {
+          parameter: arg,
+        },
+      );
+    }
+
+    usedCliArguments[i] = true;
+
+    if (paramInfo.parameterType === ParameterType.BOOLEAN) {
+      if (
+        usedCliArguments[i + 1] === false &&
```

Comment:
Can't `i + 1` be out of bounds?

Reply:
I just checked, and it can indeed be out of bounds. If the intention is for this same if to handle the array boundary, let's make that explicit. Otherwise, it looks like a mistake. 

Note that this comment is also related: https://github.com/NomicFoundation/hardhat/pull/5282/files#r1620824303

Reply:
[Done in this commit](https://github.com/NomicFoundation/hardhat/pull/5300/commits/e3e958a3ed540e89742f3fa6af3a3e6e62f3e351)